lib/src/Uncertainty/Bayesian/Gibbs.cxx

```diff
@@ -0,0 +1,199 @@
+ *  (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "openturns/Gibbs.hxx"
+#include "openturns/RandomGenerator.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+#include "openturns/ConditionalDistribution.hxx"
+#include "openturns/Log.hxx"
+#include "openturns/SpecFunc.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+
+CLASSNAMEINIT(Gibbs)
+
+static const Factory<Gibbs> Factory_Gibbs;
+
+/* Default constructor */
+Gibbs::Gibbs()
+  : RandomVectorImplementation()
+{
+  // Nothing to do
+}
+
+
+/* Parameters constructor */
+Gibbs::Gibbs(const MetropolisHastingsCollection & samplers)
+  : RandomVectorImplementation()
+  , samplers_(samplers)
+{
+  if (!samplers.getSize())
+    throw InvalidArgumentException(HERE) << "Gibbs samplers list cannot be empty";
+
+  currentState_ = samplers_[0].getImplementation()->currentState_;
+  for (UnsignedInteger j = 0; j < samplers_.getSize(); ++ j)
+  {
+    if (samplers[j].getDimension() != getDimension())
+      throw InvalidArgumentException(HERE) << "Sampler dimension is not compatible";
+
+    // ensure same initial states
+    if (j > 0)
+      samplers_[j].getImplementation()->currentState_ = currentState_;
+
+    // perform burnin/thinning only at the global level
+    samplers_[j].getImplementation()->setBurnIn(0);
+    samplers_[j].getImplementation()->setThinning(1);
+  }
+  setDescription(samplers_[0].getImplementation()->getDescription());
+}
+
+
+/* String converter */
+String Gibbs::__repr__() const
+{
+  return OSS() << "class=" << Gibbs::GetClassName()
+         << " name=" << getName()
+         << " derived from " << RandomVectorImplementation::__repr__()
+         << " samplers=" << samplers_;
+}
+
+
+Gibbs* Gibbs::clone() const
+{
+  return new Gibbs(*this);
+}
+
+UnsignedInteger Gibbs::getDimension() const
+{
+  return currentState_.getDimension();
+}
+
+Point Gibbs::getRealization() const
+{
+  // perform burnin if necessary
+  const UnsignedInteger size = getThinning() + ((samplesNumber_ < getBurnIn()) ? getBurnIn() : 0);
+
+  // check the first likelihood
+  if (samplesNumber_ == 0)
+  {
+    for (UnsignedInteger j = 0; j < samplers_.getSize(); ++ j)
+    {
+      const Scalar currentLogPosterior = samplers_[j].computeLogPosterior(samplers_[j].getImplementation()->currentState_);
+      if (currentLogPosterior <= SpecFunc::LowestScalar)
+        throw InvalidArgumentException(HERE) << "The initial state should have non-zero posterior probability density";
+    }
+  }
+
+  // see which components will need to recompute the posterior
```

Comment:
There is a problem with the way `recomputeLogPosterior_` works  in `Gibbs`. Here is a script to show the issue:

```
import openturns as ot

initial_state = [0.0, 0.0, 20.0]
target = ot.Normal(3)
weird_target = ot.ComposedDistribution([ot.Normal(), ot.Normal(), ot.Dirac(20.0)])

normal0_rwmh = ot.RandomWalkMetropolisHastings(target, initial_state, ot.Uniform(-10, 10), [0]) # samples from Normal(0,1)
normal1_rwmh = ot.RandomWalkMetropolisHastings(target, initial_state, ot.Uniform(-10, 10), [1]) # samples from Normal(0,1)
dirac_rwmh = ot.RandomWalkMetropolisHastings(weird_target, initial_state, ot.Normal(), [2])     # samples from Dirac(20)


gibbs = ot.Gibbs([normal0_rwmh, normal1_rwmh, dirac_rwmh]) # samples from Normal(0,1) x Normal(0,1) x Dirac(20)
gibbs.setThinning(1) # workaround: the default value of Gibbs thinning is 0, which is invalid
sample = gibbs.getSample(1000)
```

With the current state of the PR, all Points in `sample` are [0, 0, 20].

With the code changes I proposed in my last 2 suggestions from the previous review, the marginals 0 and 1 of sample are consistent with the standard normal distribution,
and the last marginal still always takes the value 20.

Reply:
indeed, that's a bug